sound/soc/sof/ipc4-pcm.c

```diff
@@ -59,19 +59,119 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
 }
 EXPORT_SYMBOL(sof_ipc4_set_pipeline_state);
 
+static void
+sof_ipc4_add_pipeline_to_trigger_list(struct snd_sof_dev *sdev, int state,
+				      snd_pcm_state_t pcm_state,
+				      struct snd_sof_pipeline *spipe,
+				      struct ipc4_pipeline_set_state_data *data)
+{
+	struct snd_sof_widget *pipe_widget = spipe->pipe_widget;
+	struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+
+	if (pipeline->skip_during_fe_trigger)
+		return;
+
+	switch (state) {
+	case SOF_IPC4_PIPE_RUNNING:
+		/* Do not repeat RUNNING if the pipeline is already running */
+		if (spipe->started_count == spipe->paused_count)
+			data->pipeline_ids[data->count++] = pipe_widget->instance_id;
```

Comment:
I wasn't able to understand what is done here. Specifically my concern is out-of-bounds access if count is increased beyond the pipeline_ids size.

Reply:
that would never happen. We allocate the side for data->pipeline_ids based on the pipeline_list->count already.